pennylane/pulse/transmon.py

```diff
@@ -215,3 +217,202 @@ def __add__(self, other):
+    and :math:`nu`, respectively, and can all be either fixed numbers (``float``) or depend on time
+    (``callable``).
+
+    For more realistic simulations, one may restrict the amplitude to at most
+    :math:`\Omega_{\text{max}} = 20 \text{MHz}` and the carrier frequency to deviate at most
+    :math:`\nu - \omega = \pm 1 \text{GHz}` from the qubit frequency :math:`\omega`
+    (see :func:`~.transmon_interaction` and `arXiv:2203.06818 <https://arxiv.org/abs/2203.06818>`_).
+    The phase :math:`\phi(t)` is typically a slow function of time compared to :math:`\Omega(t)`.
+
+    .. note:: Currently only supporting ``d=2`` with qudit support planned in the future.
+
+    .. seealso::
+
+        :func:`~.drive`, :func:`~.rydberg_drive`, :func:`~.transmon_interaction`
+
+    Args:
+        omega (Union[float, list[float]]): List of dressed qubit frequencies in GHz. Needs to match the length of ``wires``.
+            When passing a single float all qubits are assumed to have that same frequency.
+        connections (list[tuple(int)]): List of connections ``(i, j)`` between qubits i and j.
+            When the wires in ``connections`` are not contained in ``wires``, a warning is raised.
+        g (Union[float, list[float]]): List of coupling strengths in GHz. Needs to match the length of ``connections``.
+            When passing a single float need explicit ``wires``.
+        anharmonicity (Union[float, list[float]]): List of anharmonicities in GHz. Ignored when ``d=2``.
+            When passing a single float all qubits are assumed to have that same anharmonicity.
+        wires (Union[int, list[int]]): Label of the qubit that the drive acts upon. Can be a list of multiple wires.
+        d (int): Local Hilbert space dimension. Defaults to ``d=2`` and is currently the only supported value.
+
+    Returns:
+        HardwareHamiltonian: a :class:`~.ParametrizedHamiltonian` representing the transmon interaction
+
+    **Example**
+
+    TODO: simple catch-all example
+
+    TODO: simple integration example with transmon_interaction
+
+    """
+    if d != 2:
+        raise NotImplementedError(
+            "Currently only supporting qubits. Qutrits and qudits are planned in the future."
+        )
+
+    wires = Wires(wires)
+
+    # TODO: use sigma+ and sigma- (not necessary as terms are the same, but for consistency)
+    # We compute the `coeffs` and `observables` of the EM field
+    coeffs = [
+        AmplitudeAndPhaseAndFreq(qml.math.cos, amplitude, phase, freq),
+        AmplitudeAndPhaseAndFreq(qml.math.sin, amplitude, phase, freq),
+    ]
+
+    drive_x_term = 0.5 * sum(qml.PauliX(wire) for wire in wires)
+    drive_y_term = -0.5 * sum(qml.PauliY(wire) for wire in wires)
+
+    observables = [drive_x_term, drive_y_term]
+
+    return HardwareHamiltonian(coeffs, observables, reorder_fn=_reorder_AmpPhaseFreq)
+
+
+# pylint:disable = too-few-public-methods
+class AmplitudeAndPhaseAndFreq:
```

Comment:
I understand now that we want to check for `isinstance(..., AmplitudeAndPhaseAndFreq)`.
So I think it is fine to keep it as is, but one could consider defining the `callable_amp_and...` functions globally and check for equality with them instead. This would also allow not to check for `sum(is_callables)` but to branch into the different cases directly, based on the function equality :)

Reply:
The bigger point is to check in `_reorder_AmpPhaseFreq` for the attribute that tells what is callable and what is not callable

Reply:
Note that this is not trying to reinvent the wheel but is an extension of what is already done for `drive` and `rydberg_drive` with callable amplitude and phase https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/pulse/hardware_hamiltonian.py#L440

Reply:
Fair enough. Then it is good from my end for consistency reasons. I think it might be nicer to do this without a class also because of our ongoing move away from statefulness. 
Checking which arguments were callable is not necessary if one can check for the function name that includes that information :)

Reply:
So then we would have to replace `if isinstance(coeff, AmplitudeAndPhaseAndFreq):` in the reorder function to check that the `coeff` is callable, and if so, that its `__name__` corresponds to the one we set. Sounds a bit hacky as well 🤔 

All the info that we are checking here is static and not runtime value, isnt that okay?

Reply:
I played around with it, the changes became so big that I branched out. You can check it here, though I am not convinced yet https://github.com/PennyLaneAI/pennylane/pull/4010